### PR TITLE
(PDB-1440) implement sum, avg, min, max

### PR DESCRIPTION
**Pull request by wkalt**
Expands our function support.

### CODE REVIEW

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 1]
```diff
@@ -1175,6 +1180,14 @@
             (map->InExpression {:column (columns->fields query-rec (utils/vector-maybe column))
                                 :subquery (user-node->plan-node query-rec subquery-expression)})
 
+            [["extract" [["function" (f :guard supported-fns) column]] expr]]
+            (let [column (if (= "value" column)
```
Comment (kbarber) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Perhaps detecting column type == :multi instead is going to be more future proof here?

File: documentation/api/query/v4/facts.markdown [hunk 2]
```diff
@@ -100,10 +100,10 @@ Subquery against `/fact-contents` to get all remotely-authenticated trusted fact
 Use `count` and `group_by` to get tallies of each operating system in your
 infrastructure:
 
-    curl -X GET http://localhost:8080/pdb/query/v4/facts --data-urlencode \
-      'query=["extract",[["function","count"],"value"],
-               ["=","name","operatingsystem"],
-               ["group_by", "value"]]'
+    curl -X GET http://localhost:8080/pdb/query/v4/facts \
+    --data-urlencode 'query=["extract",[["function","count"],"value"],
```
Comment (kbarber) [inline documentation/api/query/v4/facts.markdown]: *small niggle* the other examples include a 2 space indent on the second line, I guess it makes me feel more aware this is a multi-line example.

File: documentation/api/query/v4/facts.markdown [hunk 3]
```diff
@@ -113,6 +113,22 @@ infrastructure:
         "count" : 33
     } ]
 
+Use the `avg` function and the `group_by` operator to get the average uptime
+for your nodes by environment:
+
+    curl -X GET http://localhost:8080/pdb/query/v4/facts \
+    --data-urlencode 'query=["extract", ["environment",["function","avg","value"]],
```
Comment (kbarber) [inline documentation/api/query/v4/facts.markdown]: ditto

**Comment by kbarber** [comment]
Functional tests for the examples seem fine, and working on value seems fine. When I specify a field that is non numeric, I get this:

```
2015-07-31 14:32:45,135 DEBUG [qtp811219331-93 - /pdb/query/v4/facts] [p.p.q.engine] (first (hcore/format (honeysql-from-query query)))
=> "SELECT env.name AS environment, max(name) AS \"max\" FROM factsets fs INNER JOIN facts f ON fs.id = f.factset_id INNER JOIN fact_values fv ON f.fact_value_id = fv.id INNER JOIN fact_paths fp ON f.fact_path_id = fp.id INNER JOIN value_types vt ON vt.id = fv.value_type_id LEFT JOIN environments env ON fs.environment_id = env.id WHERE (fp.depth = 0 AND ((fp.name = ?) AND ((fs.certname) in  ( SELECT certnames.certname AS certname FROM certnames LEFT JOIN catalogs ON certnames.certname = catalogs.certname LEFT JOIN factsets fs ON certnames.certname = fs.certname LEFT JOIN reports ON (certnames.certname = reports.certname AND (reports.id in (SELECT latest_report_id FROM certnames))) LEFT JOIN environments catalog_environment ON catalog_environment.id = catalogs.environment_id LEFT JOIN environments facts_environment ON facts_environment.id = fs.environment_id LEFT JOIN environments reports_environment ON reports_environment.id = reports.environment_id WHERE (certnames.deactivated IS NULL AND certnames.expired IS NULL) ) ))) GROUP BY environment"
2015-07-31 14:32:45,136 DEBUG [clojure-agent-send-off-pool-3] [c.j.b.PreparedStatementHandle] SELECT env.name AS environment, max(name) AS "max" FROM factsets fs INNER JOIN facts f ON fs.id = f.factset_id INNER JOIN fact_values fv ON f.fact_value_id = fv.id INNER JOIN fact_paths fp ON f.fact_path_id = fp.id INNER JOIN value_types vt ON vt.id = fv.value_type_id LEFT JOIN environments env ON fs.environment_id = env.id WHERE (fp.depth = 0 AND ((fp.name = 'operatingsystem') AND ((fs.certname) in  ( SELECT certnames.certname AS certname FROM certnames LEFT JOIN catalogs ON certnames.certname = catalogs.certname LEFT JOIN factsets fs ON certnames.certname = fs.certname LEFT JOIN reports ON (certnames.certname = reports.certname AND (reports.id in (SELECT latest_report_id FROM certnames))) LEFT JOIN environments catalog_environment ON catalog_environment.id = catalogs.environment_id LEFT JOIN environments facts_environment ON facts_environment.id = fs.environment_id LEFT JOIN environments reports_environment ON reports_environment.id = reports.environment_id WHERE (certnames.deactivated IS NULL AND certnames.expired IS NULL) ) ))) GROUP BY environment
2015-07-31 14:32:45,147 WARN  [qtp811219331-93 - /pdb/query/v4/facts] [o.e.j.s.HttpChannel] /pdb/query/v4/facts
javax.servlet.ServletException: org.postgresql.util.PSQLException: ERROR: column reference "name" is ambiguous
  Position: 37
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:130) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:95) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:159) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.Server.handle(Server.java:497) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310) ~[jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257) [jetty-server-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540) [jetty-io-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) [jetty-util-9.2.10.v20150310.jar:9.2.10.v20150310]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) [jetty-util-9.2.10.v20150310.jar:9.2.10.v20150310]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_55]
```

Not sure if we can guard from this earlier or not?

**Comment by wkalt** [comment]
@kbarber the issue there isn't with the functions on string-valued columns, it's that the "name" column of environments collides with the "name" field of the facts response. I've addressed this by renaming environments.name to environments.environment.

This seems like a pretty lame solution, but I think it's consistent with what we've previously done with the old names column of certnames. As long as we don't name more things "name", or create queries that combine resource parameter names and fact names, I think this will leave us in good enough shape. In the latter case I'm sure this won't be the only issue we'll need to deal with.

File: test/puppetlabs/puppetdb/http/facts_test.clj [hunk 4]
```diff
@@ -1401,11 +1434,25 @@
...
+
+     ["extract" [["function" "max" "name"] "environment"] ["~" "certname" ".*"]
+      ["group_by" "environment"]]
+     [{:environment "DEV" :max "uptime_seconds"}]}))
 
 (deftestseq structured-fact-queries
   [[version endpoint] facts-endpoints]
-  ( let [facts1 {"my_structured_fact" {"a" 1
+  (let [facts1 {"my_structured_fact" {"a" 1
                                        "b" 3.14
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/http/facts_test.clj]: alignment

**Pull request merged**